### PR TITLE
Apply to offer

### LINKED ISSUE #18: Apply for Offer - API
- Validate that user is auth and verified
- Validate that the offer exists and not accepted
- Create a new record in user_offer table and with status "New".

### PR DESCRIPTION
**Pull request by RanaEmad**
Add apply to offer functionality

### CODE REVIEW

**Review by aminetiyal — APPROVED** [review]

File: app/Services/CreateOfferUserService.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            return new ErrorResource(Response::HTTP_UNPROCESSABLE_ENTITY, "This offer has already been approved.");
+        }
+
+        $data = [
+            "user_id" => $request->user()->id,
+            "offer_id" => $offer->id
+        ];
+        $data = $this->setStatusId($data);
+        try {
+            $applicationId = $this->offerUserRepository->create($data)->id;
```
Comment (mnourrhan) [inline app/Services/CreateOfferUserService.php:38]: @RanaEmad Please use updateOrCreate for handling any error that may occur if the request hits twice.

File: app/Services/CreateOfferUserService.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            return new SuccessResource(Response::HTTP_CREATED, "Application created successfully.", ["id" => $applicationId]);
+        } catch (Exception $e) {
+            return new ErrorResource(Response::HTTP_INTERNAL_SERVER_ERROR, "Internal Server Error");
+        }
+    }
+
+    protected function setStatusId($data)
```
Comment (mnourrhan) [inline app/Services/CreateOfferUserService.php:45]: @RanaEmad Can we move this method to the status repository as it is repeated in this request and the create offer request.

File: database/migrations/2020_08_25_192340_create_offer_user_table.php [hunk 3]
```diff
@@ -13,7 +13,7 @@ class CreateOfferUserTable extends Migration
      */
     public function up()
     {
-        Schema::create('offer_user', function (Blueprint $table) {
+        Schema::create('offer_users', function (Blueprint $table) {
```
Comment (mnourrhan) [inline database/migrations/2020_08_25_192340_create_offer_user_table.php:16]: @RanaEmad  I think the common convention for naming the pivot table is using the singular name of each table. Please check the below links and let me know if I miss understood anything.
https://webdevetc.com/blog/laravel-naming-conventions
https://stackoverflow.com/questions/34897444/naming-tables-in-many-to-many-relationships-laravel/34897727

Comment (RanaEmad) [reply database/migrations/2020_08_25_192340_create_offer_user_table.php:16]: The model was not updated with $table variable so it was breaking the code, I didn't know if it was on purpose or not so I adjusted it.

Comment (mnourrhan) [reply database/migrations/2020_08_25_192340_create_offer_user_table.php:16]: @RanaEmad It is supposed to update the $table variable into the model, so please do that as I think the migrations already applied on the server so we cant update it using this way. If you need to update the table name please create a new migration file instead.

**Review by mnourrhan — APPROVED** [review]

**Pull request merged**